### ISSUE TITLE
Make `KeepExpression` infer set size if fixed?

### ISSUE THREAD

**Issue by HighDiceRoller** [HighDiceRoller/icepool#145]
A tricky thing here is that, if the expression is not bound, the size (if fixed) is not known until runtime.

**Comment by HighDiceRoller** [comment]
One approach would be to implement some sort of `initial_state` that gets the pools, or at least their sizes, as arguments. This is kind of spooky, but maybe it being only for expressions is not too bad?

**Comment by HighDiceRoller** [comment]
Another issue is determining iteration order. We would need to determine if the multiset size is known first, only after which we would know whether the keep expression can go in any order or only the one order. Even then, during the iteration we still need to know which order we actually chose.

**Comment by HighDiceRoller** [comment]
Added `KeepGenerator`.
